From 43b6a0f590c7eedd6b2cc2aa9e5e8ed643d821f8 Mon Sep 17 00:00:00 2001
From: wuxiliang <wuxiliang1@contributor>


Date: Thu, 28 Aug 2025 15:39:06 +0800
Subject: [PATCH] USERSTORY PGO/LTO/BOLT pipeline adaptation

Issue:
------
he current automated packaging pipeline does not include the PGO-optimized version.

Solution:
---------
Adapt the release workflow so that the release pipeline, Fuxi, and access control can carry the PGO version package.
---
 .gitattributes                                |   1 +
 BUILD/build.sh                                | 100 +++-
 CMakeLists.txt                                |  26 +
 local_ci/script/build_package.sh              |  34 +-
 mysql-test/std_data/tpcc10.zip                |   3 +
 .../suite/pgo/include/sysbench_cleanup.inc    |  27 ++
 .../suite/pgo/include/sysbench_prepare.inc    |  50 ++
 mysql-test/suite/pgo/include/sysbench_run.inc | 101 ++++
 mysql-test/suite/pgo/include/tpcc_cleanup.inc |  30 ++
 mysql-test/suite/pgo/include/tpcc_prepare.inc |  75 +++
 mysql-test/suite/pgo/include/tpcc_run.inc     |  97 ++++
 mysql-test/suite/pgo/my.cnf                   | 174 +++++++
 .../suite/pgo/r/sysbench_oltp_rw.result       |  18 +
 mysql-test/suite/pgo/r/tpcc_oltp.result       | 454 ++++++++++++++++++
 mysql-test/suite/pgo/t/sysbench_oltp_rw.test  |  63 +++
 mysql-test/suite/pgo/t/tpcc_oltp.test         |  17 +
 16 files changed, 1264 insertions(+), 6 deletions(-)
 create mode 100644 mysql-test/std_data/tpcc10.zip
 create mode 100644 mysql-test/suite/pgo/include/sysbench_cleanup.inc
 create mode 100644 mysql-test/suite/pgo/include/sysbench_prepare.inc
 create mode 100644 mysql-test/suite/pgo/include/sysbench_run.inc
 create mode 100644 mysql-test/suite/pgo/include/tpcc_cleanup.inc
 create mode 100644 mysql-test/suite/pgo/include/tpcc_prepare.inc
 create mode 100644 mysql-test/suite/pgo/include/tpcc_run.inc
 create mode 100644 mysql-test/suite/pgo/my.cnf
 create mode 100644 mysql-test/suite/pgo/r/sysbench_oltp_rw.result
 create mode 100644 mysql-test/suite/pgo/r/tpcc_oltp.result
 create mode 100644 mysql-test/suite/pgo/t/sysbench_oltp_rw.test
 create mode 100644 mysql-test/suite/pgo/t/tpcc_oltp.test

diff --git a/.gitattributes b/.gitattributes
index bd8f9d27909..b31a9bdff11 100644
--- a/.gitattributes
+++ b/.gitattributes
@@ -67,3 +67,4 @@ build/runtime_output_directory/xclient_unit_tests filter=lfs diff=lfs merge=lfs
 storage/dstore/dstore/euleros2.5_x86_64/lib/debug/libdstore.so filter=lfs diff=lfs merge=lfs -text
 build/archive_output_directory/libsql_main.a filter=lfs diff=lfs merge=lfs -text
 build/sql/libsql_dd.a filter=lfs diff=lfs merge=lfs -text
+mysql-test/std_data/tpcc10.zip filter=lfs diff=lfs merge=lfs -text
diff --git a/BUILD/build.sh b/BUILD/build.sh
index ffd97814373..cb33839c491 100755
--- a/BUILD/build.sh
+++ b/BUILD/build.sh
@@ -42,6 +42,9 @@ BUILD_WITH_GCOV=
 BUILD_WITH_ASAN=
 BUILD_WITH_LTO=
 BUILD_WITH_PGO=
+BUILD_PGO_SYSBENCH=
+BUILD_PGO_TPCC=
+BUILD_PGO_MTR=
 BUILD_PGO_TRAIN_SCRIPT=
 BUILD_PGO_FIRST_STAGE_ONLY=
 BUILD_WITH_BOLT=
@@ -80,8 +83,12 @@ platform=`arch`
 echo "platform=$platform"
 if [ "$platform" == "aarch64" ]; then
     dstore_platform=euleros2.9_aarch64
+    export LD_LIBRARY_PATH=$GCC_HOME/lib64:$GCC_HOME/lib:$GCC_HOME/lib/gcc/aarch64-linux-gnu/10/:$LD_LIBRARY_PATH
+    export PATH=$GCC_HOME/bin:$GCC_HOME/aarch64-unknown-linux-gnu/bin/:$PATH
 elif [ "$platform" == "x86_64" ]; then
     dstore_platform=euleros2.5_x86_64
+    export LD_LIBRARY_PATH=$GCC_HOME/lib64:$GCC_HOME/lib:$GCC_HOME/lib/gcc/x86_64-linux-gnu/10/:$LD_LIBRARY_PATH
+    export PATH=$GCC_HOME/bin:$GCC_HOME/x86_64-pc-linux-gnu/bin/:$PATH
 fi
 dstore_root_path=${BUILD_SRC_DIR}/../dstore
 gaussdb_kernel_openssl=${dstore_root_path}/GaussDBKernel-third_party_binarylibs/${dstore_platform}/kernel/dependency/openssl/comm
@@ -286,7 +293,7 @@ clean_dir()
     then
         die "Refusing to clean directory $dir"
     fi
-    #rm -rf "${dir:?}/"{*,.*} 2>/dev/null
+    rm -rf "${dir:?}/"{*,.*} 2>/dev/null
 }
 
 do_validate_src()
@@ -406,7 +413,15 @@ do_pgo_training()
         pushd mysql-test >/dev/null
         # we must run all mtr tests sequentially, otherwise the profile might
         # become a mess.
-        $mtr --suite=pgo --parallel=1 || die "$mtr $?"
+        if [ -n "$BUILD_PGO_SYSBENCH" ]
+        then
+            $mtr --suite=pgo --parallel=1 --nowarnings --testcase-timeout=300 sysbench_oltp_rw || die "$mtr $?"
+        fi
+
+        if [ -n "$BUILD_PGO_TPCC" ]
+        then
+            $mtr --suite=pgo --parallel=1 --nowarnings --testcase-timeout=300 tpcc_oltp || die "$mtr $?"
+        fi
         popd >/dev/null
     fi
 
@@ -417,6 +432,61 @@ do_pgo_training()
     fi
 }
 
+do_bolt_training()
+{
+    if ! command -v llvm-bolt &> /dev/null || ! command -v merge-fdata &> /dev/null
+    then
+        echo "============ llvm-bolt or merge-fdata not support ========="
+        return 0
+    fi
+    DEST_BOLT_DIR=$BUILD_RESULT_DIR/bolt_tmp
+    clean_dir "$DEST_BOLT_DIR"
+
+    echo "Use BOLT instrumentation for profile generation"
+
+    echo "Instrumenting ${BUILD_BUILD_DIR}/bin/mysqld"
+    mv -f ${BUILD_BUILD_DIR}/bin/mysqld ${BUILD_BUILD_DIR}/bin/mysqld.orig
+    llvm-bolt --instrument \
+      ${BUILD_BUILD_DIR}/bin/mysqld.orig -o \
+      ${BUILD_BUILD_DIR}/bin/mysqld \
+      --instrumentation-file=${DEST_BOLT_DIR}/bolt.instr.profile \
+      --instrumentation-binpath=${BUILD_BUILD_DIR}/bin/mysqld
+
+    local mtr="$PERL mtr"
+    pushd mysql-test >/dev/null
+
+    if [ -n "$BUILD_PGO_SYSBENCH" ]
+    then
+        $mtr --suite=pgo --parallel=1 --nowarnings --testcase-timeout=300 sysbench_oltp_rw || die "$mtr $?"
+        mv -f ${DEST_BOLT_DIR}/bolt.instr.profile \
+            ${DEST_BOLT_DIR}/bolt.instr.profile.sysbench_oltp_rw
+    fi
+
+    if [ -n "$BUILD_PGO_TPCC" ]
+    then
+        $mtr --suite=pgo --parallel=1 --nowarnings --testcase-timeout=300 tpcc_oltp || die "$mtr $?"
+        mv -f ${DEST_BOLT_DIR}/bolt.instr.profile \
+            ${DEST_BOLT_DIR}/bolt.instr.profile.tpcc_oltp
+    fi
+
+    popd >/dev/null
+
+    echo "Merging profiles to ${DEST_BOLT_DIR}/bolt.profile.full"
+    merge-fdata \
+      ${DEST_BOLT_DIR}/bolt.instr.profile.sysbench_oltp_rw \
+      ${DEST_BOLT_DIR}/bolt.instr.profile.tpcc_oltp > \
+      ${DEST_BOLT_DIR}/bolt.profile.full
+
+    echo "Optimize with BOLT using profile ${DEST_BOLT_DIR}/bolt.profile.full"
+    llvm-bolt ${BUILD_BUILD_DIR}/bin/mysqld.orig \
+      -o ${DEST_BOLT_DIR}/mysqld -data=${DEST_BOLT_DIR}/bolt.profile.full \
+      -reorder-blocks=cache+ -split-functions=2 -reorder-functions=hfsort+ \
+      -split-all-cold -split-eh -update-debug-sections
+
+    mv -f ${DEST_BOLT_DIR}/mysqld ${BUILD_BUILD_DIR}/bin/mysqld
+    rm -f ${BUILD_BUILD_DIR}/bin/mysqld.orig
+}
+
 do_cmake()
 {
     local cmake_opt_add=("${BUILD_CMAKE_OPTS_DEFAULT[@]}" "$@")
@@ -582,7 +652,7 @@ do_build()
 
         clean_dir "$BUILD_BUILD_DIR"
 
-        cmake_opt_add+=(-DFPROFILE_DIR="$BUILD_PGO_TRAINING_DATA" -DFPROFILE_USE=1)
+        cmake_opt_add+=(-DFPROFILE_DIR="$BUILD_PGO_TRAINING_DATA" -DFPROFILE_USE=1 -DFPROFILE_GENERATE=0 -DWITH_UNIT_TESTS=0)
     fi
 
     # LTO (for 8.0 apply for final stage only)
@@ -593,6 +663,12 @@ do_build()
 
     do_cmake "${cmake_opt_add[@]}"
     do_make
+
+    if [ -n "$BUILD_WITH_BOLT"  ] 
+    then
+        do_bolt_training
+    fi
+
     if [ -n "${BUILD_CUSTOM_INSTALL:-}" ]; then
         do_install
     fi
@@ -765,7 +841,8 @@ SHORT_OPTS="sj:r:c:Yq"
 LONG_OPTS="simulate,help,repo:,jobs:,src:,rev:,branch:,lto,pgo,pgo-train:,pgo1,\
 bolt,ci,patches:,result:,res:,cmake:,cmake-only,make-only,flags:,config:,\
 docker:,no-devtoolset,devtoolset:,strip,may-clone,diff,march:,debug,release,\
-workdir:,install:,curl,scc,gcov,quiet,silent,asan,package,exclude-test,enable-install,dstore"
+workdir:,install:,curl,scc,gcov,quiet,silent,asan,package,exclude-test,enable-install,dstore, \
+pgo_sysbench,pgo_tpcc,pgo_mtr"
 
 ARGS=$(getopt -o "$SHORT_OPTS" --long "$LONG_OPTS" -n "$0" -- "$@")
 if [ $? -ne 0 ]
@@ -827,6 +904,21 @@ while true; do
             shift
             continue
         ;;
+        '--pgo_sysbench')
+            BUILD_PGO_SYSBENCH=1
+            shift
+            continue
+        ;;
+        '--pgo_tpcc')
+            BUILD_PGO_TPCC=1
+            shift
+            continue
+        ;;
+        '--pgo_mtr')
+            BUILD_PGO_MTR=1
+            shift
+            continue
+        ;;
         '--pgo-train')
             BUILD_WITH_PGO=1
             BUILD_PGO_TRAIN_SCRIPT="$(normalize_path $2)"
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 4cf8afe802c..3c436fa87b6 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1719,6 +1719,27 @@ IF(WITH_LTO)
   ENDIF()
 ENDIF(WITH_LTO)
 
+IF (WITH_BOLT)
+  # BOLT requires linking with --emit-relocs
+  SET(CMAKE_C_LINK_FLAGS "${CMAKE_C_LINK_FLAGS} -Wl,--emit-relocs")
+  SET(CMAKE_CXX_LINK_FLAGS "${CMAKE_CXX_LINK_FLAGS} -Wl,--emit-relocs")
+
+  STRING(TOUPPER "${CMAKE_BUILD_TYPE}" FLAGS_SUFFIX)
+  IF (CMAKE_COMPILER_IS_GNUCC AND NOT CMAKE_C_COMPILER_VERSION VERSION_LESS 8)
+    SET(CMAKE_C_FLAGS_${FLAGS_SUFFIX}
+        "${CMAKE_C_FLAGS_${FLAGS_SUFFIX}} -fno-reorder-blocks-and-partition")
+    SET(CMAKE_C_FLAGS
+        "${CMAKE_C_FLAGS} -fno-reorder-blocks-and-partition")
+  ENDIF()
+
+  IF (CMAKE_COMPILER_IS_GNUCXX AND NOT CMAKE_CXX_COMPILER_VERSION VERSION_LESS 8)
+    SET(CMAKE_CXX_FLAGS_${FLAGS_SUFFIX}
+        "${CMAKE_CXX_FLAGS_${FLAGS_SUFFIX}} -fno-reorder-blocks-and-partition")
+    SET(CMAKE_CXX_FLAGS
+        "${CMAKE_CXX_FLAGS} -fno-reorder-blocks-and-partition")
+  ENDIF()
+ENDIF(WITH_BOLT)
+
 IF(APPLE)
   OPTION(WITH_DEVELOPER_ENTITLEMENTS "Add developer entitlements" OFF)
 ENDIF()
@@ -2712,6 +2733,11 @@ MESSAGE(STATUS "CMAKE_CXX_LINK_FLAGS: ${CMAKE_CXX_LINK_FLAGS}")
 MESSAGE(STATUS "CMAKE_EXE_LINKER_FLAGS ${CMAKE_EXE_LINKER_FLAGS}")
 MESSAGE(STATUS "CMAKE_MODULE_LINKER_FLAGS ${CMAKE_MODULE_LINKER_FLAGS}")
 MESSAGE(STATUS "CMAKE_SHARED_LINKER_FLAGS ${CMAKE_SHARED_LINKER_FLAGS}")
+IF(WITH_LTO)
+  MESSAGE(STATUS "CMAKE_NM: ${CMAKE_NM}")
+  MESSAGE(STATUS "CMAKE_AR: ${CMAKE_AR}")
+  MESSAGE(STATUS "CMAKE_RANLIB: ${CMAKE_RANLIB}")
+ENDIF()
 
 # If user provided WITH_NDBCLUSTER keep it in cache
 IF(HAVE_WITH_NDBCLUSTER)
diff --git a/local_ci/script/build_package.sh b/local_ci/script/build_package.sh
index d43fd4820c1..44adecb9e2c 100644
--- a/local_ci/script/build_package.sh
+++ b/local_ci/script/build_package.sh
@@ -5,6 +5,13 @@ build_type=$1
 with_asan=$2
 with_cov=${3:-0}
 with_dstore=${4:-1}
+with_pgo=${5:-0}
+with_lto=${6:-0}
+with_bolt=${7:-0}
+train_sysbench=${8:-1}
+train_tpcc=${9:-1}
+train_mtr=${10:-1}
+
 source ./common_conf.sh
 asan_option=""
 if [ "$with_asan"x == "1"x ]; then
@@ -22,12 +29,35 @@ if [ "$with_dstore"x == "1"x ]; then
   cd ${mysql_root_path}/local_ci/script
   sh build_utils.sh ${build_type}
 fi
+opt_option=()
+if [[ "$with_pgo" == "1" ]];then
+  opt_option+=("--pgo")
+fi
+if [[ "$with_lto" == "1" ]];then
+  opt_option+=("--lto")
+fi
+if [[ "$with_bolt" == "1" ]];then
+  opt_option+=("--bolt")
+fi
+train_option=()
+if [[ "$train_sysbench" == "1" ]];then
+  train_option+=("--pgo_sysbench")
+fi
+if [[ "$train_tpcc" == "1" ]];then
+  train_option+=("--pgo_tpcc")
+fi
+if [[ "$train_mtr" == "1" ]];then
+  train_option+=("--pgo_mtr")
+fi
+
+echo ${opt_option[@]}
+echo ${train_option[@]}
 
 cd ${mysql_root_path}
 if [[ "${build_type}" = "debug" ]]; then
-  sh +x ${mysql_root_path}/BUILD/build_debug.sh  --workdir=${mysql_root_path}/hwsql-builder --result=${mysql_root_path}/artis --install=${install_path}/mysql --jobs=$(nproc) --enable-install $asan_option $cov_option $dstore_option
+  sh +x ${mysql_root_path}/BUILD/build_debug.sh  --workdir=${mysql_root_path}/hwsql-builder --result=${mysql_root_path}/artis --install=${install_path}/mysql --jobs=$(nproc) --enable-install $asan_option $cov_option $dstore_option ${opt_option[@]} ${train_option[@]}
 else
-  sh +x ${mysql_root_path}/BUILD/build_release.sh  --workdir=${mysql_root_path}/hwsql-builder --result=${mysql_root_path}/artis --install=${install_path}/mysql --jobs=$(nproc) --enable-install $asan_option $cov_option $dstore_option
+  sh +x ${mysql_root_path}/BUILD/build_release.sh  --workdir=${mysql_root_path}/hwsql-builder --result=${mysql_root_path}/artis --install=${install_path}/mysql --jobs=$(nproc) --enable-install $asan_option $cov_option $dstore_option ${opt_option[@]} ${train_option[@]}
 fi
 if [ $? -ne 0 ];then
   echo "Compile failed"
diff --git a/mysql-test/std_data/tpcc10.zip b/mysql-test/std_data/tpcc10.zip
new file mode 100644
index 00000000000..281c203cad0
--- /dev/null
+++ b/mysql-test/std_data/tpcc10.zip
@@ -0,0 +1,3 @@
+version https://git-lfs.github.com/spec/v1
+oid sha256:f9ce350b07f31fecfc17d910c95cb1b3bad18bcc615d3132b5babbd8c9a92ae8
+size 450542341
diff --git a/mysql-test/suite/pgo/include/sysbench_cleanup.inc b/mysql-test/suite/pgo/include/sysbench_cleanup.inc
new file mode 100644
index 00000000000..f1eef6ca5cc
--- /dev/null
+++ b/mysql-test/suite/pgo/include/sysbench_cleanup.inc
@@ -0,0 +1,27 @@
+########################################################################
+#
+# SUMMARY
+#
+#   Drop tables created by sysbench_prepare.inc
+#
+# USAGE
+#
+#   let $tables=10;
+#   --source suite/pgo/include/sysbench_cleanup.inc
+#
+#   $tables defaults to 1.
+#
+########################################################################
+
+if (!$tables)
+{
+  let $tables = 1;
+}
+
+--disable_query_log
+while ($tables)
+{
+  eval DROP TABLE sbtest$tables;
+  dec $tables;
+}
+--enable_query_log
diff --git a/mysql-test/suite/pgo/include/sysbench_prepare.inc b/mysql-test/suite/pgo/include/sysbench_prepare.inc
new file mode 100644
index 00000000000..c84d02bed0d
--- /dev/null
+++ b/mysql-test/suite/pgo/include/sysbench_prepare.inc
@@ -0,0 +1,50 @@
+########################################################################
+#
+# SUMMARY
+#
+#   Create and populate table(s) using the sysbench OLTP schema
+#
+# USAGE
+#
+#   let $tables=10;
+#   --source suite/pgo/include/sysbench_prepare.inc
+#
+#   $tables defaults to 1.
+#
+########################################################################
+
+if (!$tables)
+{
+  let $tables = 1;
+}
+
+--disable_query_log
+while ($tables)
+{
+  eval CREATE TABLE sbtest$tables(
+    id INTEGER NOT NULL AUTO_INCREMENT,
+    k INTEGER DEFAULT '0' NOT NULL,
+    c CHAR(120) DEFAULT '' NOT NULL,
+    pad CHAR(60) DEFAULT '' NOT NULL,
+    PRIMARY KEY (id)) ENGINE=Dstore;
+  eval INSERT INTO sbtest$tables(id, k, c, pad) VALUES
+    (1, FLOOR(1 + RAND() * 1048576),
+    CONCAT(REPEAT(CONCAT(FLOOR(RAND()*100000000000), '-'), 9), FLOOR(RAND()*100000000000)),
+    CONCAT(REPEAT(CONCAT(FLOOR(RAND()*100000000000), '-'), 4), FLOOR(RAND()*100000000000)));
+  # Generate 2^20 = 1048576 rows
+  let $i = 0;
+  let $base = 1;
+  while ($i < 20)
+  {
+    eval INSERT INTO sbtest$tables(id, k, c, pad)
+      SELECT $base + id, FLOOR(1 + RAND() * 1048576), c, pad FROM sbtest$tables;
+    inc $i;
+    let $base = $base * 2;
+  }
+
+  eval CREATE INDEX k_$tables ON sbtest$tables(k);
+  eval ANALYZE TABLE sbtest$tables;
+
+  dec $tables;
+}
+--enable_query_log
diff --git a/mysql-test/suite/pgo/include/sysbench_run.inc b/mysql-test/suite/pgo/include/sysbench_run.inc
new file mode 100644
index 00000000000..bfde4c4b29e
--- /dev/null
+++ b/mysql-test/suite/pgo/include/sysbench_run.inc
@@ -0,0 +1,101 @@
+
+##############################################################################
+#
+# SUMMARY
+#
+#   Run sysbench OLTP RW benchmark on table(s) created by sysbench_prepare.inc
+#
+# USAGE
+#
+#   let $tables=10;
+#   let transactions=10000
+#   --source suite/pgo/include/sysbench_run.inc
+#
+#   $tables defaults to 1.
+#
+#   $transactions defaults to 500 (should take about 30 seconds on a
+#   PGO-enabled build).
+#
+##############################################################################
+
+if (!$tables)
+{
+  let $tables = 1;
+}
+
+if (!$transactions)
+{
+  let $transactions = 500;
+}
+
+--disable_query_log
+--disable_result_log
+
+while ($transactions)
+{
+  START TRANSACTION;
+
+  # Point SELECTs
+  let $table=`SELECT FLOOR(1 + (RAND() * $tables))`;
+  let $pscount=10;
+  while ($pscount)
+  {
+    let $id=`SELECT FLOOR(1 + RAND() * 1048576)`;
+    eval SELECT * FROM sbtest$table WHERE id=$id;
+    dec $pscount;
+  }
+
+  # Simple range
+  let $table=`SELECT FLOOR(1 + (RAND() * $tables))`;
+  let $id=`SELECT FLOOR(1 + RAND() * 1048576)`;
+  eval SELECT c FROM sbtest$table WHERE id BETWEEN $id and $id+100;
+
+  # Sum range
+  let $table=`SELECT FLOOR(1 + (RAND() * $tables))`;
+  let $id=`SELECT FLOOR(1 + RAND() * 1048576)`;
+  eval SELECT SUM(k) FROM sbtest$table WHERE id BETWEEN $id and $id+100;
+
+  # Order range
+  let $table=`SELECT FLOOR(1 + (RAND() * $tables))`;
+  let $id=`SELECT FLOOR(1 + RAND() * 1048576)`;
+  eval SELECT c FROM sbtest$table WHERE id BETWEEN $id and $id+100 ORDER BY c;
+
+  # Distinct range
+  let $table=`SELECT FLOOR(1 + (RAND() * $tables))`;
+  let $id=`SELECT FLOOR(1 + RAND() * 1048576)`;
+  eval SELECT DISTINCT c FROM sbtest$table WHERE id BETWEEN $id and $id+100 ORDER BY c;
+
+  # Index update
+  let $table=`SELECT FLOOR(1 + (RAND() * $tables))`;
+  let $id=`SELECT FLOOR(1 + RAND() * 1048576)`;
+  eval UPDATE sbtest$table SET k=k+1 WHERE id=$id;
+
+  # Non-index update
+  let $table=`SELECT FLOOR(1 + (RAND() * $tables))`;
+  let $id=`SELECT FLOOR(1 + RAND() * 1048576)`;
+  eval UPDATE sbtest$table
+    SET c=CONCAT(REPEAT(CONCAT(FLOOR(RAND()*100000000000), '-'), 9), FLOOR(RAND()*100000000000))
+    WHERE id=$id;
+
+  # DELETE
+  let $table=`SELECT FLOOR(1 + (RAND() * $tables))`;
+  let $id=`SELECT FLOOR(1 + RAND() * 1048576)`;
+  eval DELETE FROM sbtest$table WHERE id=$id;
+
+  # INSERT
+  eval INSERT INTO sbtest$table VALUES
+    ($id, FLOOR(1 + RAND() * 1048576),
+    CONCAT(REPEAT(CONCAT(FLOOR(RAND()*100000000000), '-'), 9), FLOOR(RAND()*100000000000)),
+    CONCAT(REPEAT(CONCAT(FLOOR(RAND()*100000000000), '-'), 4), FLOOR(RAND()*100000000000)));
+
+  COMMIT;
+
+  dec $transactions;
+}
+
+# Reset transactions, so we don't use its current value (0) on repeated
+# inclusions
+let $transactions=;
+
+--enable_result_log
+--enable_query_log
diff --git a/mysql-test/suite/pgo/include/tpcc_cleanup.inc b/mysql-test/suite/pgo/include/tpcc_cleanup.inc
new file mode 100644
index 00000000000..e453fc47ff8
--- /dev/null
+++ b/mysql-test/suite/pgo/include/tpcc_cleanup.inc
@@ -0,0 +1,30 @@
+########################################################################
+#
+# SUMMARY
+#
+#   Drop tables created by tpcc_prepare.inc
+#
+# USAGE
+#
+#   --source suite/pgo/include/tpcc_cleanup.inc
+########################################################################
+
+SET @OLD_UNIQUE_CHECKS=@@UNIQUE_CHECKS, UNIQUE_CHECKS=0;
+SET @OLD_FOREIGN_KEY_CHECKS=@@FOREIGN_KEY_CHECKS, FOREIGN_KEY_CHECKS=0;
+
+--disable_query_log
+# Drop tpcc tables
+DROP TABLE customer;
+DROP TABLE district;
+DROP TABLE history;
+DROP TABLE item;
+DROP TABLE new_orders;
+DROP TABLE order_line;
+DROP TABLE orders;
+DROP TABLE stock;
+DROP TABLE warehouse;
+DROP DATABASE tpcc10;
+--enable_query_log
+
+SET FOREIGN_KEY_CHECKS=@OLD_FOREIGN_KEY_CHECKS;
+SET UNIQUE_CHECKS=@OLD_UNIQUE_CHECKS;
diff --git a/mysql-test/suite/pgo/include/tpcc_prepare.inc b/mysql-test/suite/pgo/include/tpcc_prepare.inc
new file mode 100644
index 00000000000..7117b3c629f
--- /dev/null
+++ b/mysql-test/suite/pgo/include/tpcc_prepare.inc
@@ -0,0 +1,75 @@
+########################################################################
+#
+# SUMMARY
+#
+#   Create and populate table(s) using the TPC-C schema
+#
+# USAGE
+#   --source suite/pgo/include/tpcc_prepare.inc
+#
+########################################################################
+
+--disable_query_log
+--disable_result_log
+--disable_abort_on_error
+
+--exec unzip -qo $MYSQLTEST_VARDIR/std_data/tpcc10.zip -d $MYSQL_TMP_DIR
+
+SET @OLD_UNIQUE_CHECKS=@@UNIQUE_CHECKS, UNIQUE_CHECKS=0;
+SET @OLD_FOREIGN_KEY_CHECKS=@@FOREIGN_KEY_CHECKS, FOREIGN_KEY_CHECKS=0;
+
+DROP DATABASE IF EXISTS tpcc10;
+
+CREATE DATABASE tpcc10;
+USE tpcc10;
+--echo # load tpcc data
+--exec $MYSQL --show-warnings --user=root tpcc10 < $MYSQL_TMP_DIR/tpcc10/customer.sql
+--exec $MYSQL --show-warnings --user=root tpcc10 < $MYSQL_TMP_DIR/tpcc10/district.sql
+--exec $MYSQL --show-warnings --user=root tpcc10 < $MYSQL_TMP_DIR/tpcc10/history.sql
+--exec $MYSQL --show-warnings --user=root tpcc10 < $MYSQL_TMP_DIR/tpcc10/item.sql
+--exec $MYSQL --show-warnings --user=root tpcc10 < $MYSQL_TMP_DIR/tpcc10/new_orders.sql
+--exec $MYSQL --show-warnings --user=root tpcc10 < $MYSQL_TMP_DIR/tpcc10/order_line.sql
+--exec $MYSQL --show-warnings --user=root tpcc10 < $MYSQL_TMP_DIR/tpcc10/orders.sql
+--exec $MYSQL --show-warnings --user=root tpcc10 < $MYSQL_TMP_DIR/tpcc10/stock.sql
+--exec $MYSQL --show-warnings --user=root tpcc10 < $MYSQL_TMP_DIR/tpcc10/warehouse.sql
+
+--exec rm -rf $MYSQL_TMP_DIR/tpcc10
+
+SET FOREIGN_KEY_CHECKS=@OLD_FOREIGN_KEY_CHECKS;
+SET UNIQUE_CHECKS=@OLD_UNIQUE_CHECKS;
+
+--echo # analyze tables
+ANALYZE TABLE customer;
+ANALYZE TABLE district;
+ANALYZE TABLE history;
+ANALYZE TABLE item;
+ANALYZE TABLE new_orders;
+ANALYZE TABLE order_line;
+ANALYZE TABLE orders;
+ANALYZE TABLE stock;
+ANALYZE TABLE warehouse;
+
+--enable_result_log
+--enable_query_log
+--enable_abort_on_error
+
+--echo # check table
+SHOW CREATE TABLE customer;
+SHOW CREATE TABLE district;
+SHOW CREATE TABLE history;
+SHOW CREATE TABLE item;
+SHOW CREATE TABLE new_orders;
+SHOW CREATE TABLE order_line;
+SHOW CREATE TABLE orders;
+SHOW CREATE TABLE stock;
+SHOW CREATE TABLE warehouse;
+
+SELECT COUNT(*) FROM customer;
+SELECT COUNT(*) FROM district;
+SELECT COUNT(*) FROM history;
+SELECT COUNT(*) FROM item;
+SELECT COUNT(*) FROM new_orders;
+SELECT COUNT(*) FROM order_line;
+SELECT COUNT(*) FROM orders;
+SELECT COUNT(*) FROM stock;
+SELECT COUNT(*) FROM warehouse;
diff --git a/mysql-test/suite/pgo/include/tpcc_run.inc b/mysql-test/suite/pgo/include/tpcc_run.inc
new file mode 100644
index 00000000000..7ed0dd8b7b1
--- /dev/null
+++ b/mysql-test/suite/pgo/include/tpcc_run.inc
@@ -0,0 +1,97 @@
+##############################################################################
+#
+# SUMMARY
+#
+#   Run TPC-C test on table(s) created by tpcc_prepare.inc
+#
+# USAGE
+#
+#   let loop=10
+#   --source suite/pgo/include/tpcc_run.inc
+#
+##############################################################################
+
+USE tpcc10;
+
+if (!$loop)
+{
+  let $loop = 50;
+}
+
+--disable_query_log
+--disable_result_log
+
+--echo loop : $loop
+while ($loop)
+{
+    let $warehouse_id = `SELECT FLOOR(1 + RAND() * 10)`;
+    let $district_id = `SELECT FLOOR(1 + RAND() * 10)`;
+    let $customer_id = `SELECT FLOOR(1 + RAND() * 3000)`;
+    let $item_id = `SELECT FLOOR(1 + RAND() * 100000)`;
+    let $order_id = `SELECT FLOOR(3000 + RAND() * 10000)`;
+
+    --echo # do new order
+    BEGIN;
+    --eval SELECT d_next_o_id, d_tax FROM district WHERE d_id = $district_id AND d_w_id = $warehouse_id FOR UPDATE;
+    --eval UPDATE district SET d_next_o_id = d_next_o_id + 1 WHERE d_id = $district_id AND d_w_id = $warehouse_id
+    --eval INSERT INTO orders (o_id, o_d_id, o_w_id, o_c_id, o_entry_d, o_carrier_id, o_ol_cnt, o_all_local) VALUES ($order_id, $district_id, $warehouse_id, $customer_id, NOW(), NULL, 5, 1);
+    --eval INSERT INTO new_orders (no_o_id, no_d_id, no_w_id) VALUES ($order_id, $district_id, $warehouse_id);
+    --eval SELECT i_price, i_name, i_data FROM item WHERE i_id = $item_id;
+    --eval SELECT s_quantity, s_data, s_dist_01, s_dist_02, s_dist_03, s_dist_04, s_dist_05, s_dist_06, s_dist_07, s_dist_08, s_dist_09, s_dist_10 FROM stock WHERE s_i_id = $item_id AND s_w_id = $warehouse_id FOR UPDATE;
+    # insert order_line
+    let $i = 1;
+    while ($i <= 5) {
+        --eval INSERT INTO order_line (ol_o_id, ol_d_id, ol_w_id, ol_number, ol_i_id, ol_supply_w_id, ol_quantity, ol_amount, ol_dist_info) VALUES ($order_id, $district_id, $warehouse_id, $i, $item_id, $warehouse_id, 5, 10.50, 'dist_info')
+        inc $i;
+        inc $item_id;
+    }
+    # update stock
+    --eval UPDATE stock SET s_quantity = s_quantity - 5 WHERE s_i_id = $item_id AND s_w_id = $warehouse_id;
+    COMMIT;
+
+    --echo # do payment
+    BEGIN;
+    --eval UPDATE warehouse SET w_ytd = w_ytd + 100 WHERE w_id = $warehouse_id;
+    --eval UPDATE district SET d_ytd = d_ytd + 100 WHERE d_id = $district_id AND d_w_id = $warehouse_id;
+    --eval SELECT count(c_id) FROM customer WHERE c_w_id = $warehouse_id AND c_d_id = $district_id;
+    --eval SELECT c_first, c_middle, c_last, c_street_1, c_street_2, c_city, c_state, c_zip, c_phone, c_credit, c_credit_lim, c_discount, c_balance, c_since FROM customer WHERE c_w_id = $warehouse_id AND c_d_id = $district_id AND c_id = $customer_id FOR UPDATE;
+    --eval UPDATE customer SET c_balance = c_balance - 100, c_ytd_payment = c_ytd_payment + 100 WHERE c_id = $customer_id AND c_d_id = $district_id AND c_w_id = $warehouse_id;
+    --eval INSERT INTO history (h_c_id, h_c_d_id, h_c_w_id, h_d_id, h_w_id, h_date, h_amount, h_data) VALUES ($customer_id, $district_id, $warehouse_id, $district_id, $warehouse_id, NOW(), 100, 'Payment');
+    COMMIT;
+
+    --echo # do order stat
+    BEGIN;
+    --eval SELECT count(c_id) FROM customer WHERE c_w_id = $warehouse_id AND c_d_id = $district_id AND c_last = "ABLEABLEANTI";
+    --eval SELECT c_balance, c_first, c_middle, c_last FROM customer WHERE c_w_id = $warehouse_id AND c_d_id = $district_id AND c_last = "ABLEABLEANTI" ORDER BY c_first;
+    --eval SELECT o_id, o_carrier_id, o_entry_d FROM orders WHERE o_w_id = $warehouse_id AND o_d_id = $district_id AND o_c_id = $customer_id ORDER BY o_id DESC LIMIT 1;
+    --eval SELECT o_id, o_entry_d, COALESCE(o_carrier_id,0) FROM orders WHERE o_w_id = $warehouse_id AND o_d_id = $district_id AND o_c_id = $customer_id AND o_id = (SELECT MAX(o_id) FROM orders WHERE o_w_id = $warehouse_id AND o_d_id = $district_id AND o_c_id = $customer_id);
+    --eval SELECT ol_i_id, ol_supply_w_id, ol_quantity, ol_amount, ol_delivery_d FROM order_line WHERE ol_w_id = $warehouse_id AND ol_d_id = $district_id AND ol_o_id = $order_id;
+    COMMIT;
+
+    --echo # do delivery
+    BEGIN;
+    --eval SELECT COALESCE(MIN(no_o_id),0) FROM new_orders WHERE no_d_id = $district_id AND no_w_id = $warehouse_id;
+    --eval SELECT no_o_id INTO @order_id FROM new_orders WHERE no_w_id = $warehouse_id AND no_d_id = $district_id ORDER BY no_o_id ASC LIMIT 1;
+    --eval UPDATE orders SET o_carrier_id = 1 WHERE o_id = $order_id AND o_d_id = $district_id AND o_w_id = $warehouse_id;
+    --eval UPDATE order_line SET ol_delivery_d = NOW() WHERE ol_o_id = $order_id AND ol_d_id = $district_id AND ol_w_id = $warehouse_id;
+    --eval DELETE FROM new_orders WHERE no_o_id = $order_id AND no_d_id = $district_id AND no_w_id = $warehouse_id;
+    --eval SELECT SUM(ol_amount) FROM order_line WHERE ol_o_id = $order_id AND ol_d_id = $district_id AND ol_w_id = $warehouse_id;
+    --eval UPDATE customer SET c_balance = c_balance + 100 , c_delivery_cnt = c_delivery_cnt + 1 WHERE c_id = $customer_id AND c_d_id = $district_id AND c_w_id = $warehouse_id;
+    COMMIT;
+
+    --echo # do stock level
+    BEGIN;
+    --eval SELECT d_next_o_id FROM district WHERE d_id = $district_id AND d_w_id = $warehouse_id;
+    --eval SELECT DISTINCT ol_i_id FROM order_line WHERE ol_w_id = $warehouse_id AND ol_d_id = $district_id AND ol_o_id < $order_id AND ol_o_id >= 1500;
+    --eval SELECT COUNT(*) FROM stock WHERE s_w_id = $warehouse_id AND s_i_id = $item_id AND s_quantity < 10;
+    COMMIT;
+
+    dec $loop;
+}
+
+--enable_result_log
+--enable_query_log
+
+--echo # check data
+--eval SELECT COUNT(*) AS orphaned_new_orders FROM new_orders no LEFT JOIN orders o ON no.no_o_id = o.o_id AND no.no_d_id = o.o_d_id AND no.no_w_id = o.o_w_id WHERE o.o_id IS NULL;
+--eval SELECT COUNT(*) AS inconsistent_stock FROM stock WHERE s_quantity < 0;
diff --git a/mysql-test/suite/pgo/my.cnf b/mysql-test/suite/pgo/my.cnf
new file mode 100644
index 00000000000..8ab17fc3bcc
--- /dev/null
+++ b/mysql-test/suite/pgo/my.cnf
@@ -0,0 +1,174 @@
+# my.cnf for the PGO training suite. This should be as close as
+# possible to my.cnf used on production systems, with the exception of
+# resource consumption, because resources on build slaves are usually
+# limited. That is why buffer sizes are left at their default values.
+
+!include include/default_client.cnf
+
+[mysqld.1]
+
+connect-timeout=60
+
+# When binary logging is enabled, this variable controls whether
+# stored function creators can be trusted not to create stored
+# functions that will cause unsafe events to be written to the binary log.
+# Set it to 1 so that no restrictions are enforced on
+# stored function creation.
+log-bin-trust-function-creators=1
+
+# Running MTR with --mem causes var directories to be created in memory.
+# Set innodb-buffer-pool-size and innodb-redo-log-capacity to a lower
+# value because running tests in parallel, on the ramdisk may consume
+# excessive memory.
+innodb-buffer-pool-size=24M
+innodb-redo-log-capacity=10M
+# MAINTAINER:
+# the loose- syntax is to make sure the cnf file is also
+# valid when building without the performance schema.
+
+# Run tests with the performance schema instrumentation
+loose-enable-performance-schema
+# Run tests with a small number of instrumented objects
+# to limit memory consumption with MTR
+loose-performance-schema-accounts-size=100
+loose-performance-schema-digests-size=200
+loose-performance-schema-hosts-size=100
+loose-performance-schema-users-size=100
+loose-performance-schema-setup-actors-size=100
+loose-performance-schema-setup-objects-size=100
+
+loose-performance-schema-events-waits-history-size=10
+loose-performance-schema-events-waits-history-long-size=10000
+loose-performance-schema-events-stages-history-size=10
+loose-performance-schema-events-stages-history-long-size=1000
+loose-performance-schema-events-statements-history-size=10
+loose-performance-schema-events-statements-history-long-size=1000
+loose-performance-schema-events-transactions-history-size=10
+loose-performance-schema-events-transactions-history-long-size=1000
+loose-performance-schema-session-connect-attrs-size=2048
+
+# Enable everything, for maximun code exposure during testing
+
+loose-performance-schema-instrument='%=ON'
+
+loose-performance-schema-consumer-events-stages-current=ON
+loose-performance-schema-consumer-events-stages-history=ON
+loose-performance-schema-consumer-events-stages-history-long=ON
+loose-performance-schema-consumer-events-statements-cpu=ON
+loose-performance-schema-consumer-events-statements-current=ON
+loose-performance-schema-consumer-events-statements-history=ON
+loose-performance-schema-consumer-events-statements-history-long=ON
+loose-performance-schema-consumer-events-transactions-current=ON
+loose-performance-schema-consumer-events-transactions-history=ON
+loose-performance-schema-consumer-events-transactions-history-long=ON
+loose-performance-schema-consumer-events-waits-current=ON
+loose-performance-schema-consumer-events-waits-history=ON
+loose-performance-schema-consumer-events-waits-history-long=ON
+loose-performance-schema-consumer-global-instrumentation=ON
+loose-performance-schema-consumer-thread-instrumentation=ON
+
+# Set LOG_ERROR_VERBOSITY to 3 as default for MTR tests. This will
+# make sure that even note messeges are getting logged in mysqld error
+# log file. The note messages will contain important information and
+# will be useful while debugging an issue as well.
+log-error-verbosity=3
+storage_engine_mode=BOTH_STORAGE_ENGINE
+rds_permission_control=OFF
+loose-rds_restricted_dynamic_privileges_control=OFF
+loose-rds_extended_partitions_enabled = ON
+loose-rds_disable_user_defined_functions = OFF
+
+bind-address=0.0.0.0
+binlog-ignore-db=information_schema
+binlog_format=ROW
+character_set_server=UTF8MB4
+default_storage_engine=Dstore
+innodb_stats_include_delete_marked=ON
+enforce_gtid_consistency=ON
+expire_logs_days=1
+explicit_defaults_for_timestamp=OFF
+gtid_mode=on
+
+innodb_file_per_table=ON
+innodb_flush_method=O_DIRECT
+
+log_bin_trust_function_creators=ON
+log_slave_updates=true
+
+master_info_repository=TABLE
+max_binlog_size=67108864
+
+relay-log=rds_mysql-relay-bin
+replicate_ignore_db=information_schema
+
+skip-name-resolve
+
+slave_parallel_workers=2
+slow_query_log=ON
+
+sql_mode=NO_ENGINE_SUBSTITUTION,STRICT_ALL_TABLES
+sync_binlog=1
+
+slave_parallel_type=LOGICAL_CLOCK
+performance_schema=OFF
+
+block_encryption_mode=aes-128-cbc
+binlog_error_action=ABORT_SERVER
+
+relay_log_info_repository=FILE
+local_infile=ON
+skip_external_locking=ON
+max_allowed_packet=1073741824
+
+innodb_flush_neighbors=0
+innodb_io_capacity=12000
+innodb_io_capacity_max=24000
+
+innodb_page_cleaners=16
+
+innodb_adaptive_hash_index=1
+
+threadpool_enabled=ON
+threadpool_size=128
+
+transaction_isolation=READ-COMMITTED
+table_open_cache=16384
+table_open_cache_instances=64
+thread_cache_size=128
+
+dstore_buffer=131072
+dstore_buffer_lru_partition=256
+dstore_check_point_timeout=60
+dstore_maintenance_work_mem=65536
+dstore_ncores=2770
+dstore_log_min_messages=7
+dstore_fold_period=0
+dstore_fold_level=0
+dstore_lock_hash_table_size=28160
+dstore_lock_table_partition_num=16
+dstore_enable_lazy_lock=false
+dstore_vfs_tenant_isolation_config_path=null
+dstore_num_obj_space_mgr_workers=5
+dstore_prob_of_extension_threshold=1
+dstore_wal_level=2
+dstore_recovery_worker_num=1
+dstore_synchronous_commit=true
+dstore_wal_file_size=3221225472
+dstore_wal_buffers=131072
+dstore_wal_flush_timeout=300
+dstore_wal_read_buffer_size=536870912
+dstore_wal_redo_buffer_size=268435456
+dstore_wal_writer_cpu_bind=48
+dstore_wal_keep_segments=16
+dstore_bg_wal_writer_min_bytes=51200
+dstore_wal_each_write_lenghth_limit=524288
+dstore_disable_btree_page_recycle=false
+dstore_deadlock_time_interval=20000
+dstore_bg_disk_writer_slave_num=1
+dstore_bg_page_writer_sleep_millisecond=2000
+dstore_wal_throttling_size=838860800
+dstore_max_io_capacity_kb=1048576
+dstore_bulk_read_ring_size=16384
+dstore_bulk_write_ring_size=2097152
+dstore_csn_thread_bind_cpu=33
+dstore_enable_perf_timers=false
diff --git a/mysql-test/suite/pgo/r/sysbench_oltp_rw.result b/mysql-test/suite/pgo/r/sysbench_oltp_rw.result
new file mode 100644
index 00000000000..aa1025a95f4
--- /dev/null
+++ b/mysql-test/suite/pgo/r/sysbench_oltp_rw.result
@@ -0,0 +1,18 @@
+call mtr.add_suppression("'NO_ZERO_DATE', 'NO_ZERO_IN_DATE' and 'ERROR_FOR_DIVISION_BY_ZERO' sql modes should be used with strict mode");
+SELECT @@dstore_buffer;
+@@dstore_buffer
+131072
+Table	Op	Msg_type	Msg_text
+test.sbtest1	analyze	status	OK
+SELECT COUNT(*), MIN(id), MAX(id) FROM sbtest1;
+COUNT(*)	MIN(id)	MAX(id)
+1048576	1	1048576
+CREATE USER root@127.0.0.1;
+GRANT ALL PRIVILEGES ON *.* TO root@127.0.0.1;
+SHOW STATUS LIKE 'Ssl_cipher';
+Variable_name	Value
+Ssl_cipher	SSL_CIPHER
+SHOW STATUS LIKE 'Ssl_cipher';
+Variable_name	Value
+Ssl_cipher	SSL_CIPHER
+DROP USER root@127.0.0.1;
diff --git a/mysql-test/suite/pgo/r/tpcc_oltp.result b/mysql-test/suite/pgo/r/tpcc_oltp.result
new file mode 100644
index 00000000000..eccd780f7fd
--- /dev/null
+++ b/mysql-test/suite/pgo/r/tpcc_oltp.result
@@ -0,0 +1,454 @@
+SELECT @@dstore_buffer;
+@@dstore_buffer
+131072
+SELECT @@sql_mode;
+@@sql_mode
+STRICT_ALL_TABLES,NO_ENGINE_SUBSTITUTION
+SELECT @@performance_schema;
+@@performance_schema
+0
+# load tpcc data
+# analyze tables
+# check table
+SHOW CREATE TABLE customer;
+Table	Create Table
+customer	CREATE TABLE `customer` (
+  `c_id` int NOT NULL,
+  `c_d_id` tinyint NOT NULL,
+  `c_w_id` smallint NOT NULL,
+  `c_first` varchar(16) DEFAULT NULL,
+  `c_middle` char(2) DEFAULT NULL,
+  `c_last` varchar(16) DEFAULT NULL,
+  `c_street_1` varchar(20) DEFAULT NULL,
+  `c_street_2` varchar(20) DEFAULT NULL,
+  `c_city` varchar(20) DEFAULT NULL,
+  `c_state` char(2) DEFAULT NULL,
+  `c_zip` char(9) DEFAULT NULL,
+  `c_phone` char(16) DEFAULT NULL,
+  `c_since` datetime DEFAULT NULL,
+  `c_credit` char(2) DEFAULT NULL,
+  `c_credit_lim` bigint DEFAULT NULL,
+  `c_discount` decimal(4,2) DEFAULT NULL,
+  `c_balance` decimal(12,2) DEFAULT NULL,
+  `c_ytd_payment` decimal(12,2) DEFAULT NULL,
+  `c_payment_cnt` smallint DEFAULT NULL,
+  `c_delivery_cnt` smallint DEFAULT NULL,
+  `c_data` varchar(8192) DEFAULT NULL,
+  PRIMARY KEY (`c_w_id`,`c_d_id`,`c_id`),
+  KEY `idx_customer` (`c_w_id`,`c_d_id`,`c_last`,`c_first`,`c_id`),
+  CONSTRAINT `fkey_customer_1` FOREIGN KEY (`c_w_id`, `c_d_id`) REFERENCES `district` (`d_w_id`, `d_id`)
+) ENGINE=Dstore DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
+SHOW CREATE TABLE district;
+Table	Create Table
+district	CREATE TABLE `district` (
+  `d_id` tinyint NOT NULL,
+  `d_w_id` smallint NOT NULL,
+  `d_name` varchar(10) DEFAULT NULL,
+  `d_street_1` varchar(20) DEFAULT NULL,
+  `d_street_2` varchar(20) DEFAULT NULL,
+  `d_city` varchar(20) DEFAULT NULL,
+  `d_state` char(2) DEFAULT NULL,
+  `d_zip` char(9) DEFAULT NULL,
+  `d_tax` decimal(4,2) DEFAULT NULL,
+  `d_ytd` decimal(12,2) DEFAULT NULL,
+  `d_next_o_id` int DEFAULT NULL,
+  PRIMARY KEY (`d_w_id`,`d_id`),
+  CONSTRAINT `fkey_district_1` FOREIGN KEY (`d_w_id`) REFERENCES `warehouse` (`w_id`)
+) ENGINE=Dstore DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
+SHOW CREATE TABLE history;
+Table	Create Table
+history	CREATE TABLE `history` (
+  `h_c_id` int DEFAULT NULL,
+  `h_c_d_id` tinyint DEFAULT NULL,
+  `h_c_w_id` smallint DEFAULT NULL,
+  `h_d_id` tinyint DEFAULT NULL,
+  `h_w_id` smallint DEFAULT NULL,
+  `h_date` datetime DEFAULT NULL,
+  `h_amount` decimal(6,2) DEFAULT NULL,
+  `h_data` varchar(24) DEFAULT NULL,
+  KEY `fkey_history_1` (`h_c_w_id`,`h_c_d_id`,`h_c_id`),
+  KEY `fkey_history_2` (`h_w_id`,`h_d_id`),
+  CONSTRAINT `fkey_history_1` FOREIGN KEY (`h_c_w_id`, `h_c_d_id`, `h_c_id`) REFERENCES `customer` (`c_w_id`, `c_d_id`, `c_id`),
+  CONSTRAINT `fkey_history_2` FOREIGN KEY (`h_w_id`, `h_d_id`) REFERENCES `district` (`d_w_id`, `d_id`)
+) ENGINE=Dstore DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
+SHOW CREATE TABLE item;
+Table	Create Table
+item	CREATE TABLE `item` (
+  `i_id` int NOT NULL,
+  `i_im_id` int DEFAULT NULL,
+  `i_name` varchar(24) DEFAULT NULL,
+  `i_price` decimal(5,2) DEFAULT NULL,
+  `i_data` varchar(50) DEFAULT NULL,
+  PRIMARY KEY (`i_id`)
+) ENGINE=Dstore DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
+SHOW CREATE TABLE new_orders;
+Table	Create Table
+new_orders	CREATE TABLE `new_orders` (
+  `no_o_id` int NOT NULL,
+  `no_d_id` tinyint NOT NULL,
+  `no_w_id` smallint NOT NULL,
+  PRIMARY KEY (`no_w_id`,`no_d_id`,`no_o_id`),
+  CONSTRAINT `fkey_new_orders_1` FOREIGN KEY (`no_w_id`, `no_d_id`, `no_o_id`) REFERENCES `orders` (`o_w_id`, `o_d_id`, `o_id`)
+) ENGINE=Dstore DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
+SHOW CREATE TABLE order_line;
+Table	Create Table
+order_line	CREATE TABLE `order_line` (
+  `ol_o_id` int NOT NULL,
+  `ol_d_id` tinyint NOT NULL,
+  `ol_w_id` smallint NOT NULL,
+  `ol_number` tinyint NOT NULL,
+  `ol_i_id` int DEFAULT NULL,
+  `ol_supply_w_id` smallint DEFAULT NULL,
+  `ol_delivery_d` datetime DEFAULT NULL,
+  `ol_quantity` tinyint DEFAULT NULL,
+  `ol_amount` decimal(6,2) DEFAULT NULL,
+  `ol_dist_info` char(24) DEFAULT NULL,
+  PRIMARY KEY (`ol_w_id`,`ol_d_id`,`ol_o_id`,`ol_number`),
+  KEY `fkey_order_line_2` (`ol_supply_w_id`,`ol_i_id`),
+  CONSTRAINT `fkey_order_line_1` FOREIGN KEY (`ol_w_id`, `ol_d_id`, `ol_o_id`) REFERENCES `orders` (`o_w_id`, `o_d_id`, `o_id`),
+  CONSTRAINT `fkey_order_line_2` FOREIGN KEY (`ol_supply_w_id`, `ol_i_id`) REFERENCES `stock` (`s_w_id`, `s_i_id`)
+) ENGINE=Dstore DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
+SHOW CREATE TABLE orders;
+Table	Create Table
+orders	CREATE TABLE `orders` (
+  `o_id` int NOT NULL,
+  `o_d_id` tinyint NOT NULL,
+  `o_w_id` smallint NOT NULL,
+  `o_c_id` int DEFAULT NULL,
+  `o_entry_d` datetime DEFAULT NULL,
+  `o_carrier_id` tinyint DEFAULT NULL,
+  `o_ol_cnt` tinyint DEFAULT NULL,
+  `o_all_local` tinyint DEFAULT NULL,
+  PRIMARY KEY (`o_w_id`,`o_d_id`,`o_id`),
+  KEY `bmsql_oorder_idx1` (`o_w_id`,`o_d_id`,`o_c_id`,`o_id` DESC),
+  CONSTRAINT `fkey_orders_1` FOREIGN KEY (`o_w_id`, `o_d_id`, `o_c_id`) REFERENCES `customer` (`c_w_id`, `c_d_id`, `c_id`)
+) ENGINE=Dstore DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
+SHOW CREATE TABLE stock;
+Table	Create Table
+stock	CREATE TABLE `stock` (
+  `s_i_id` int NOT NULL,
+  `s_w_id` smallint NOT NULL,
+  `s_quantity` smallint DEFAULT NULL,
+  `s_dist_01` char(24) DEFAULT NULL,
+  `s_dist_02` char(24) DEFAULT NULL,
+  `s_dist_03` char(24) DEFAULT NULL,
+  `s_dist_04` char(24) DEFAULT NULL,
+  `s_dist_05` char(24) DEFAULT NULL,
+  `s_dist_06` char(24) DEFAULT NULL,
+  `s_dist_07` char(24) DEFAULT NULL,
+  `s_dist_08` char(24) DEFAULT NULL,
+  `s_dist_09` char(24) DEFAULT NULL,
+  `s_dist_10` char(24) DEFAULT NULL,
+  `s_ytd` decimal(8,0) DEFAULT NULL,
+  `s_order_cnt` smallint DEFAULT NULL,
+  `s_remote_cnt` smallint DEFAULT NULL,
+  `s_data` varchar(50) DEFAULT NULL,
+  PRIMARY KEY (`s_w_id`,`s_i_id`),
+  KEY `fkey_stock_2` (`s_i_id`),
+  CONSTRAINT `fkey_stock_1` FOREIGN KEY (`s_w_id`) REFERENCES `warehouse` (`w_id`),
+  CONSTRAINT `fkey_stock_2` FOREIGN KEY (`s_i_id`) REFERENCES `item` (`i_id`)
+) ENGINE=Dstore DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
+SHOW CREATE TABLE warehouse;
+Table	Create Table
+warehouse	CREATE TABLE `warehouse` (
+  `w_id` smallint NOT NULL,
+  `w_name` varchar(10) DEFAULT NULL,
+  `w_street_1` varchar(20) DEFAULT NULL,
+  `w_street_2` varchar(20) DEFAULT NULL,
+  `w_city` varchar(20) DEFAULT NULL,
+  `w_state` char(2) DEFAULT NULL,
+  `w_zip` char(9) DEFAULT NULL,
+  `w_tax` decimal(4,2) DEFAULT NULL,
+  `w_ytd` decimal(12,2) DEFAULT NULL,
+  PRIMARY KEY (`w_id`)
+) ENGINE=Dstore DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
+SELECT COUNT(*) FROM customer;
+COUNT(*)
+300000
+SELECT COUNT(*) FROM district;
+COUNT(*)
+100
+SELECT COUNT(*) FROM history;
+COUNT(*)
+300000
+SELECT COUNT(*) FROM item;
+COUNT(*)
+100000
+SELECT COUNT(*) FROM new_orders;
+COUNT(*)
+90000
+SELECT COUNT(*) FROM order_line;
+COUNT(*)
+2999046
+SELECT COUNT(*) FROM orders;
+COUNT(*)
+300000
+SELECT COUNT(*) FROM stock;
+COUNT(*)
+1000000
+SELECT COUNT(*) FROM warehouse;
+COUNT(*)
+10
+USE tpcc10;
+loop : 50
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# do new order
+# do payment
+# do order stat
+# do delivery
+# do stock level
+# check data
+SELECT COUNT(*) AS orphaned_new_orders FROM new_orders no LEFT JOIN orders o ON no.no_o_id = o.o_id AND no.no_d_id = o.o_d_id AND no.no_w_id = o.o_w_id WHERE o.o_id IS NULL;;
+orphaned_new_orders
+0
+SELECT COUNT(*) AS inconsistent_stock FROM stock WHERE s_quantity < 0;;
+inconsistent_stock
+0
+SET @OLD_UNIQUE_CHECKS=@@UNIQUE_CHECKS, UNIQUE_CHECKS=0;
+SET @OLD_FOREIGN_KEY_CHECKS=@@FOREIGN_KEY_CHECKS, FOREIGN_KEY_CHECKS=0;
+SET FOREIGN_KEY_CHECKS=@OLD_FOREIGN_KEY_CHECKS;
+SET UNIQUE_CHECKS=@OLD_UNIQUE_CHECKS;
diff --git a/mysql-test/suite/pgo/t/sysbench_oltp_rw.test b/mysql-test/suite/pgo/t/sysbench_oltp_rw.test
new file mode 100644
index 00000000000..6a232247eac
--- /dev/null
+++ b/mysql-test/suite/pgo/t/sysbench_oltp_rw.test
@@ -0,0 +1,63 @@
+--source include/have_nodebug.inc
+--source include/have_openssl.inc
+
+call mtr.add_suppression("'NO_ZERO_DATE', 'NO_ZERO_IN_DATE' and 'ERROR_FOR_DIVISION_BY_ZERO' sql modes should be used with strict mode");
+
+# The default dataset size with a single sysbench table containing 1M rows is
+# ~240 MB. We deliberately want to run the workload with the default InnoDB
+# buffer pool size (128 MB) to cover code paths related to both cached and
+# non-cached data
+
+SELECT @@dstore_buffer;
+
+--source suite/pgo/include/sysbench_prepare.inc
+
+# Validate the prepared dataset
+SELECT COUNT(*), MIN(id), MAX(id) FROM sbtest1;
+
+CREATE USER root@127.0.0.1;
+GRANT ALL PRIVILEGES ON *.* TO root@127.0.0.1;
+
+# Save the initial number of concurrent sessions
+--source include/count_sessions.inc
+
+# Test with a non-SSL connection on a Unix socket
+connect (con0,localhost,root,,,,,);
+connection con0;
+--source suite/pgo/include/sysbench_run.inc
+connection default;
+disconnect con0;
+
+# Test with an SSL connection on a Unix socket
+connect (con0,localhost,root,,,,,SSL);
+connection con0;
+--replace_regex $ALLOWED_CIPHERS_REGEX
+SHOW STATUS LIKE 'Ssl_cipher';
+--source suite/pgo/include/sysbench_run.inc
+connection default;
+disconnect con0;
+
+# Test with a non-SSL connection on a TCP socket
+connect (con0,127.0.0.1,root,,,$MASTER_PORT,,);
+connection con0;
+--source suite/pgo/include/sysbench_run.inc
+connection default;
+disconnect con0;
+
+# Test with an SSL connection on a TCP socket
+connect (con0,127.0.0.1,root,,,$MASTER_PORT,,SSL);
+connection con0;
+--replace_regex $ALLOWED_CIPHERS_REGEX
+SHOW STATUS LIKE 'Ssl_cipher';
+--source suite/pgo/include/sysbench_run.inc
+connection default;
+disconnect con0;
+
+# Cleanup
+--source suite/pgo/include/sysbench_cleanup.inc
+
+# Wait till we reached the initial number of concurrent sessions
+--source include/wait_until_count_sessions.inc
+
+DROP USER root@127.0.0.1;
+
diff --git a/mysql-test/suite/pgo/t/tpcc_oltp.test b/mysql-test/suite/pgo/t/tpcc_oltp.test
new file mode 100644
index 00000000000..bc0da2cf280
--- /dev/null
+++ b/mysql-test/suite/pgo/t/tpcc_oltp.test
@@ -0,0 +1,17 @@
+--source include/have_nodebug.inc
+
+# The default dataset is TPCC with 10 warehouses.
+# We set dstore_buffer to 2GB, not all data in buffer pool.
+
+SELECT @@dstore_buffer;
+SELECT @@sql_mode;
+SELECT @@performance_schema;
+
+# Prepare tpc-c table and data for test
+--source suite/pgo/include/tpcc_prepare.inc
+
+# Run OLAP test cases
+--source suite/pgo/include/tpcc_run.inc
+
+# Cleanup
+--source suite/pgo/include/tpcc_cleanup.inc
-- 
2.23.0

